index.js

```diff
@@ -156,41 +156,65 @@ class Statham {
   /**
    * Fetches value of given key.
    *
-   * @param {String} key
-   * @param {{}}     [data] Base object to search in
+   * @param {String|Array} key
+   * @param {*}            [defaultValue] Value to return if key was not found
    *
    * @returns {*}
    */
-  fetch(key, data) {
-    let rest = Utils.normalizeKey(key);
-    key      = rest.shift();
-    data     = data || this.data;
+  fetch(key, defaultValue) {
+    defaultValue = typeof defaultValue === 'undefined' ? null : defaultValue;
+
+    if (typeof this.data[key] !== 'undefined') {
+      return this.data[key];
+    }
+
+    if (this.isModeFlat()) {
+      return defaultValue;
+    }
+
+    let keys    = Utils.normalizeKey(key);
+    let lastKey = keys.pop();
+    let tmp     = this.data;
+
+    for (let i = 0; i < keys.length; i++) {
+      if (typeof tmp[keys[i]] === 'undefined') {
+        return defaultValue;
+      }
+
+      tmp = tmp[keys[i]];
+    }
 
-    return rest.length === 0 ? data[key] : this.fetch(rest, data[key]);
+    return typeof tmp[lastKey] === 'undefined' ? defaultValue : tmp[lastKey];
```

Comment:
I might want to store undefined as a value on my object's property. Should it not return undefined in that case?